fvm/environment/uuids.go

```diff
@@ -39,45 +52,124 @@ func (generator ParseRestrictedUUIDGenerator) GenerateUUID() (uint64, error) {
 
 type uUIDGenerator struct {
 	tracer tracing.TracerSpan
+	log    zerolog.Logger
 	meter  Meter
 
 	txnState state.NestedTransactionPreparer
+
+	blockHeader *flow.Header
+	txnIndex    uint32
+
+	initialized bool
+	partition   byte
+	registerId  flow.RegisterID
+}
+
+func uuidPartition(blockId flow.Identifier, txnIndex uint32) byte {
+	// Partitioning by txnIndex ensures temporally neighboring transactions do
+	// not share registers / conflict with each other.
+	//
+	// Since all blocks will have a transaction at txnIndex 0 but not
+	// necessarily a transaction at txnIndex 255, if we assign partition based
+	// only on txnIndex, partition 0's counter (and other low-valued
+	// partitions' counters) will fill up much more quickly than high-valued
+	// partitions' counters.  Therefore, a deterministically random offset is
+	// used to ensure the partitioned counters are roughly balanced.  Any byte
+	// in the sha hash is sufficiently random/uniform for this purpose (Note that
+	// block Id is already a sha hash, but its hash implementation may change
+	// underneath us).
+	//
+	// Note that since partition 0 reuses the legacy counter, its counter is
+	// much	further ahead than the other partitions.  If partition 0's counter
+	// is in danager of overflowing, use variants of "the power of two random
+	// choices" to shed load to other counters.
```

Comment:
right, if we don't want to deal with migration, we could just start the partition from 1, and leave the first partition (partition 0) to the legacy counter. Because otherwise, the first partition will start the counter value from 0, which has already been used the legacy counter. The second partition will start the counter from 100....0000, which won't overlap with legacy counter. To ensure this, we need to do two things:
1. partition should always greater than 0
2. counter value should include the partition in the first byte

I don't find the above two logic right now. Did I miss?

Reply:
I think currently counter is at ~ 31.5 - 32 bits  (Approx 0.000003% of all space used so far for partition 0 ) , to 56 bits there is a really long long way. I hope to be wrong someday, but partition 0 will never be full for sure.

Reply:
Just found both logic. 

It’s backward compatibility, implemented by UUIDRegisterID function

Reply:
as bluesign pointed out, we can basically assume all partitions, including partition, 0 will never overflow.